lms/djangoapps/instructor/handlers.py

```diff
@@ -0,0 +1,83 @@
+"""
+Handlers for instructor
+"""
+import logging
+
+from django.contrib.auth import get_user_model
+from django.core.exceptions import ObjectDoesNotExist
+from django.dispatch import receiver
+from openedx_events.learning.signals import EXAM_ATTEMPT_RESET, EXAM_ATTEMPT_SUBMITTED
+
+from lms.djangoapps.courseware.models import StudentModule
+from lms.djangoapps.instructor import enrollment
+from lms.djangoapps.instructor.tasks import update_exam_completion_task
+
+User = get_user_model()
+
+log = logging.getLogger(__name__)
+
+
+@receiver(EXAM_ATTEMPT_SUBMITTED)
+def handle_exam_completion(sender, signal, **kwargs):
+    """
+    exam completion event from the event bus
+    """
+    event_data = kwargs.get('exam_attempt')
+    user_data = event_data.student_user
+    usage_key = event_data.usage_key
+
+    update_exam_completion_task.apply_async((user_data.pii.username, str(usage_key), 1.0))
+
+
+@receiver(EXAM_ATTEMPT_RESET)
+def handle_exam_reset(sender, signal, **kwargs):
+    """
+    exam reset event from the event bus
+    """
+    event_data = kwargs.get('exam_attempt')
+    user_data = event_data.student_user
+    requesting_user_data = event_data.requesting_user
+    usage_key = event_data.usage_key
+    course_key = event_data.course_key
+    content_id = str(usage_key)
+
+    try:
```

Comment:
I just noticed that this code is mostly a copy/paste of the `delete_student_attempt()` code in `services.py`. Should we do something similar to [what we did in the invalidate certificate consumer PR](https://github.com/openedx/edx-platform/pull/33319/files#diff-0f3ac000f80eebb6a8b7cbdae7496934f9e06a9d75fee952725d0f3d324bfbb9R927), where we compartmentalize this code in an api function (or something similar)?

Reply:
Side note that you may have to modify your handlers and services tests, and may also need to add a test to test_api.py for coverage.

Reply:
The code here is similar but I'd argue a helper function adds complexity since the purpose of these two functions is to interface with a different system with slightly different inputs.

Thinking a bit forward to the fact that services.py eventually goes away. Better to duplicate and easily delete half than need to refactor shared code later.

I think there's something to be said about where the call to reset_student_attempts is. It could probably be refactored into the `update_completion_task` but I stopped short of that to avoid reworking too many things.

Reply:
I see, in that case, should I refactor my code in the [invalidate certificate PR](https://github.com/openedx/edx-platform/pull/33319/files#diff-0f3ac000f80eebb6a8b7cbdae7496934f9e06a9d75fee952725d0f3d324bfbb9R927) to not have a helper function for the same reason?